src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamWriter.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.streaming;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.DataOutputStreamPlus;
+import org.apache.cassandra.net.async.ByteBufDataOutputStreamPlus;
+import org.apache.cassandra.streaming.ProgressInfo;
+import org.apache.cassandra.streaming.StreamManager;
+import org.apache.cassandra.streaming.StreamSession;
+import org.apache.cassandra.utils.FBUtilities;
+
+import static org.apache.cassandra.streaming.StreamManager.StreamRateLimiter;
+
+/**
+ * CassandraBlockStreamWriter streams the entire SSTable to given channel.
+ */
+public class CassandraBlockStreamWriter implements IStreamWriter
+{
+    private static final Logger logger = LoggerFactory.getLogger(CassandraBlockStreamWriter.class);
+
+    private final SSTableReader sstable;
+    private final ComponentManifest manifest;
+    private final StreamSession session;
+    private final StreamRateLimiter limiter;
+
+    public CassandraBlockStreamWriter(SSTableReader sstable, StreamSession session, ComponentManifest manifest)
+    {
+        this.session = session;
+        this.sstable = sstable;
+        this.manifest = manifest;
+        this.limiter =  StreamManager.getRateLimiter(session.peer);
+    }
+
+    /**
+     * Stream the entire file to given channel.
+     * <p>
+     *
+     * @param output where this writes data to
+     * @throws IOException on any I/O error
+     */
+    @Override
+    public void write(DataOutputStreamPlus output) throws IOException
+    {
+        long totalSize = manifest.getTotalSize();
+        logger.debug("[Stream #{}] Start streaming sstable {} to {}, repairedAt = {}, totalSize = {}", session.planId(),
+                     sstable.getFilename(), session.peer, sstable.getSSTableMetadata().repairedAt, totalSize);
+
+        long progress = 0L;
+        ByteBufDataOutputStreamPlus byteBufDataOutputStreamPlus = (ByteBufDataOutputStreamPlus) output;
+
+        for (Component component : manifest.getComponents())
+        {
+            @SuppressWarnings("resource") // this is closed after the file is transferred by ByteBufDataOutputStreamPlus
+            FileChannel in = new RandomAccessFile(sstable.descriptor.filenameFor(component), "r").getChannel();
+
+            // Total Length to transmit for this file
+            long length = in.size();
+
+            // tracks write progress
+            long bytesRead = 0;
```

Comment:
Is this a variable left from some previous iteration of the class? Because we only update it once.. perhaps should be called `bytesWritten` (I assume it was copy-pasted from the read impl?)

Reply:
I did rename this.